File: CalcYouLater/CalculatorEngine.swift
import Foundation
import Combine

struct HistoryEntry: Identifiable {
    let id = UUID()
    let expression: String
    let result: String
    let timestamp: Date
}

class CalculatorEngine: ObservableObject {
    @Published var display: String = "0"
    @Published var expression: String = ""
    @Published var history: [HistoryEntry] = []
    @Published var memory: Double = 0
    @Published var hasMemory: Bool = false

    private var leftOperand: Double?
    private var pendingOperator: String?
    private var lastOperand: Double?
    private var lastOperator: String?
    private var shouldResetDisplay = false
    private var justEvaluated = false

    var displayValue: Double { Double(display.replacingOccurrences(of: ",", with: "")) ?? 0 }

    // MARK: - Digit / Decimal Input

    func inputDigit(_ digit: String) {
        if shouldResetDisplay {
            display = digit
            shouldResetDisplay = false
            if justEvaluated {
                leftOperand = nil
                pendingOperator = nil
                lastOperand = nil
                lastOperator = nil
                justEvaluated = false
            }
        } else if display == "0" {
            display = digit
        } else {
            guard display.count < 15 else { return }
            display += digit
        }
    }

    func inputDecimal() {
        if shouldResetDisplay || justEvaluated {
            display = "0."
            shouldResetDisplay = false
            justEvaluated = false
        } else if !display.contains(".") {
            display += "."
        }
    }

    func backspace() {
        guard !justEvaluated, !shouldResetDisplay else { return }
        if display.count > 1 {
            display = String(display.dropLast())
            if display == "-" { display = "0" }
        } else {
            display = "0"
        }
    }

    // MARK: - Operators

    func inputOperator(_ op: String) {
        let current = displayValue

        if justEvaluated {
            leftOperand = current
            justEvaluated = false
        } else if let left = leftOperand, let pending = pendingOperator, !shouldResetDisplay {
            let result = compute(left, pending, current)
            addToHistory(expression: "\(fmt(left)) \(pending) \(fmt(current))", result: fmt(result))
            display = fmt(result)
            leftOperand = result
        } else {
            leftOperand = current
        }

        pendingOperator = op
        expression = "\(fmt(leftOperand!)) \(op)"
        shouldResetDisplay = true
    }

    func equals() {
        if justEvaluated {
            guard let left = leftOperand, let op = lastOperator, let right = lastOperand else { return }
            let result = compute(left, op, right)
            addToHistory(expression: "\(fmt(left)) \(op) \(fmt(right)) =", result: fmt(result))
            expression = "\(fmt(left)) \(op) \(fmt(right)) ="
            display = fmt(result)
            leftOperand = result
            return
        }

        let current = displayValue
        guard let left = leftOperand, let op = pendingOperator else {
            if let op = lastOperator, let right = lastOperand {
                let result = compute(current, op, right)
                addToHistory(expression: "\(fmt(current)) \(op) \(fmt(right)) =", result: fmt(result))
                expression = "\(fmt(current)) \(op) \(fmt(right)) ="
                display = fmt(result)
                leftOperand = result
                shouldResetDisplay = true
                justEvaluated = true
            }
            return
        }

        lastOperand = current
        lastOperator = op
        let result = compute(left, op, current)
        addToHistory(expression: "\(fmt(left)) \(op) \(fmt(current)) =", result: fmt(result))
        expression = "\(fmt(left)) \(op) \(fmt(current)) ="
        display = fmt(result)
        leftOperand = result
        pendingOperator = nil
        shouldResetDisplay = true
        justEvaluated = true
    }

    // MARK: - Clear

    func clear() {
        if display != "0" && !justEvaluated {
            display = "0"
        } else {
            allClear()
        }
    }

    func allClear() {
        display = "0"
        expression = ""
        leftOperand = nil
        pendingOperator = nil
        lastOperand = nil
        lastOperator = nil
        shouldResetDisplay = false
        justEvaluated = false
    }

    var clearLabel: String { (leftOperand == nil && !justEvaluated) || display == "0" ? "AC" : "C" }

    // MARK: - Quick Functions

    func toggleSign() {
        let val = displayValue
        display = fmt(-val)
        if justEvaluated { leftOperand = -val }
    }

    func percent() {
        let val = displayValue
        if let left = leftOperand, let _ = pendingOperator {
            display = fmt(left * val / 100)
        } else {
            display = fmt(val / 100)
        }
    }

    // MARK: - Scientific

    func applyScientific(_ fn: String) {
        let val = displayValue
        let result: Double

        switch fn {
        case "sin":   result = sin(val * .pi / 180)
        case "cos":   result = cos(val * .pi / 180)
        case "tan":   result = tan(val * .pi / 180)
        case "asin":  result = asin(val) * 180 / .pi
        case "acos":  result = acos(val) * 180 / .pi
        case "atan":  result = atan(val) * 180 / .pi
        case "log":   result = log10(val)
        case "ln":    result = log(val)
        case "sqrt":  result = sqrt(val)
        case "cbrt":  result = cbrt(val)
        case "x²":    result = val * val
        case "x³":    result = val * val * val
        case "1/x":   result = 1.0 / val
        case "n!":    result = factorial(Int(abs(val)))
        case "abs":   result = abs(val)
        case "e^x":   result = exp(val)
        case "10^x":  result = pow(10, val)
        default: return
        }

        addToHistory(expression: "\(fn)(\(fmt(val)))", result: fmt(result))
        expression = "\(fn)(\(fmt(val))) ="
        display = fmt(result)
        leftOperand = nil
        pendingOperator = nil
        shouldResetDisplay = true
        justEvaluated = true
    }

    func inputConstant(_ c: String) {
        switch c {
        case "π": display = fmt(Double.pi)
        case "e": display = fmt(M_E)
        default: return
        }
        shouldResetDisplay = true
        justEvaluated = false
    }

    // MARK: - Memory

    func memoryClear()    { memory = 0; hasMemory = false }
    func memoryRecall()   { display = fmt(memory); shouldResetDisplay = true }
    func memoryAdd()      { memory += displayValue; hasMemory = true }
    func memorySubtract() { memory -= displayValue; hasMemory = memory != 0 }

    // MARK: - History

    func recallHistory(_ entry: HistoryEntry) {
        display = entry.result
        shouldResetDisplay = true
        leftOperand = nil
        pendingOperator = nil
        expression = entry.expression
        justEvaluated = false
    }

    func clearHistory() { history.removeAll() }

    // MARK: - Formatting

    func fmt(_ value: Double) -> String {
        if value.isNaN { return "Error" }
        if value.isInfinite { return value > 0 ? "∞" : "-∞" }

        let absVal = abs(value)
        if absVal >= 1e15 || (absVal < 1e-9 && absVal > 0) {
            let f = NumberFormatter()
            f.numberStyle = .scientific
            f.maximumSignificantDigits = 8
            return f.string(from: NSNumber(value: value)) ?? "\(value)"
        }

        if value == value.rounded(.towardZero) && absVal < 1e15 {
            if let i = Int64(exactly: value) { return "\(i)" }
        }

        return String(format: "%.10g", value)
    }

    // MARK: - Private

    private func compute(_ a: Double, _ op: String, _ b: Double) -> Double {
        switch op {
        case "+":  return a + b
        case "−":  return a - b
        case "×":  return a * b
        case "÷":  return b == 0 ? (a >= 0 ? .infinity : -.infinity) : a / b
        case "xʸ": return pow(a, b)
        default:   return b
        }
    }

    private func factorial(_ n: Int) -> Double {
        guard n >= 0 else { return .nan }
        if n > 20 { return .infinity }
        return n == 0 ? 1 : Double(n) * factorial(n - 1)
    }

    private func addToHistory(expression: String, result: String) {
        guard result != "Error" else { return }
        history.insert(HistoryEntry(expression: expression, result: result, timestamp: Date()), at: 0)
        if history.count > 200 { history.removeLast() }
    }
}

File: CalcYouLater-Android/app/src/main/java/com/calcyoulater/android/engine/CalculatorEngine.kt
package com.calcyoulater.android.engine

import kotlin.math.PI
import kotlin.math.E
import kotlin.math.abs
import kotlin.math.acos
import kotlin.math.asin
import kotlin.math.atan
import kotlin.math.cbrt
import kotlin.math.cos
import kotlin.math.exp
import kotlin.math.ln
import kotlin.math.log10
import kotlin.math.pow
import kotlin.math.sin
import kotlin.math.sqrt
import kotlin.math.tan

/**
 * Immutable snapshot of engine-visible state, consumed by the UI layer.
 */
data class EngineState(
    val display: String,
    val expression: String,
    val history: List<HistoryEntry>,
    val memory: Double,
    val hasMemory: Boolean,
    val clearLabel: String
)

/**
 * Pure-Kotlin port of CalculatorEngine.swift. No Android / Compose dependencies,
 * so it is fully unit-testable on the JVM. The UI wraps this via a ViewModel and
 * reads [snapshot] after each mutating call.
 *
 * Unicode operators are preserved exactly to match macOS/iOS history strings:
 *   "−" U+2212, "×" U+00D7, "÷" U+00F7, "xʸ".
 */
class CalculatorEngine {

    var display: String = "0"; private set
    var expression: String = ""; private set
    val history: MutableList<HistoryEntry> = mutableListOf()
    var memory: Double = 0.0; private set
    var hasMemory: Boolean = false; private set

    private var leftOperand: Double? = null
    private var pendingOperator: String? = null
    private var lastOperand: Double? = null
    private var lastOperator: String? = null
    private var shouldResetDisplay = false
    private var justEvaluated = false

    private val displayValue: Double
        get() = display.replace(",", "").toDoubleOrNull() ?: 0.0

    fun snapshot(): EngineState = EngineState(
        display = display,
        expression = expression,
        history = history.toList(),
        memory = memory,
        hasMemory = hasMemory,
        clearLabel = clearLabel
    )

    // MARK: - Digit / Decimal Input

    fun inputDigit(digit: String) {
        if (shouldResetDisplay) {
            display = digit
            shouldResetDisplay = false
            if (justEvaluated) {
                leftOperand = null
                pendingOperator = null
                lastOperand = null
                lastOperator = null
                justEvaluated = false
            }
        } else if (display == "0") {
            display = digit
        } else {
            if (display.length >= 15) return
            display += digit
        }
    }

    fun inputDecimal() {
        if (shouldResetDisplay || justEvaluated) {
            display = "0."
            shouldResetDisplay = false
            justEvaluated = false
        } else if (!display.contains(".")) {
            display += "."
        }
    }

    fun backspace() {
        if (justEvaluated || shouldResetDisplay) return
        if (display.length > 1) {
            display = display.dropLast(1)
            if (display == "-") display = "0"
        } else {
            display = "0"
        }
    }

    // MARK: - Operators

    fun inputOperator(op: String) {
        val current = displayValue
        val left = leftOperand
        val pending = pendingOperator

        if (justEvaluated) {
            leftOperand = current
            justEvaluated = false
        } else if (left != null && pending != null && !shouldResetDisplay) {
            val result = compute(left, pending, current)
            addToHistory("${fmt(left)} $pending ${fmt(current)}", fmt(result))
            display = fmt(result)
            leftOperand = result
        } else {
            leftOperand = current
        }

        pendingOperator = op
        expression = "${fmt(leftOperand!!)} $op"
        shouldResetDisplay = true
    }

    fun equals() {
        if (justEvaluated) {
            val left = leftOperand ?: return
            val op = lastOperator ?: return
            val right = lastOperand ?: return
            val result = compute(left, op, right)
            addToHistory("${fmt(left)} $op ${fmt(right)} =", fmt(result))
            expression = "${fmt(left)} $op ${fmt(right)} ="
            display = fmt(result)
            leftOperand = result
            return
        }

        val current = displayValue
        val left = leftOperand
        val op = pendingOperator
        if (left == null || op == null) {
            val lop = lastOperator
            val right = lastOperand
            if (lop != null && right != null) {
                val result = compute(current, lop, right)
                addToHistory("${fmt(current)} $lop ${fmt(right)} =", fmt(result))
                expression = "${fmt(current)} $lop ${fmt(right)} ="
                display = fmt(result)
                leftOperand = result
                shouldResetDisplay = true
                justEvaluated = true
            }
            return
        }

        lastOperand = current
        lastOperator = op
        val result = compute(left, op, current)
        addToHistory("${fmt(left)} $op ${fmt(current)} =", fmt(result))
        expression = "${fmt(left)} $op ${fmt(current)} ="
        display = fmt(result)
        leftOperand = result
        pendingOperator = null
        shouldResetDisplay = true
        justEvaluated = true
    }

    // MARK: - Clear

    fun clear() {
        if (display != "0" && !justEvaluated) {
            display = "0"
        } else {
            allClear()
        }
    }

    fun allClear() {
        display = "0"
        expression = ""
        leftOperand = null
        pendingOperator = null
        lastOperand = null
        lastOperator = null
        shouldResetDisplay = false
        justEvaluated = false
    }

    val clearLabel: String
        get() = if ((leftOperand == null && !justEvaluated) || display == "0") "AC" else "C"

    // MARK: - Quick Functions

    fun toggleSign() {
        val v = displayValue
        display = fmt(-v)
        if (justEvaluated) leftOperand = -v
    }

    fun percent() {
        val v = displayValue
        val left = leftOperand
        if (left != null && pendingOperator != null) {
            display = fmt(left * v / 100)
        } else {
            display = fmt(v / 100)
        }
    }

    // MARK: - Scientific

    fun applyScientific(fn: String) {
        val v = displayValue
        val result: Double = when (fn) {
            "sin" -> sin(v * PI / 180)
            "cos" -> cos(v * PI / 180)
            "tan" -> tan(v * PI / 180)
            "asin" -> asin(v) * 180 / PI
            "acos" -> acos(v) * 180 / PI
            "atan" -> atan(v) * 180 / PI
            "log" -> log10(v)
            "ln" -> ln(v)
            "sqrt" -> sqrt(v)
            "cbrt" -> cbrt(v)
            "x²" -> v * v
            "x³" -> v * v * v
            "1/x" -> 1.0 / v
            "n!" -> factorial(abs(v).toInt())
            "abs" -> abs(v)
            "e^x" -> exp(v)
            "10^x" -> 10.0.pow(v)
            else -> return
        }

        addToHistory("$fn(${fmt(v)})", fmt(result))
        expression = "$fn(${fmt(v)}) ="
        display = fmt(result)
        leftOperand = null
        pendingOperator = null
        shouldResetDisplay = true
        justEvaluated = true
    }

    fun inputConstant(c: String) {
        when (c) {
            "π" -> display = fmt(PI)
            "e" -> display = fmt(E)
            else -> return
        }
        shouldResetDisplay = true
        justEvaluated = false
    }

    // MARK: - Memory

    fun memoryClear() { memory = 0.0; hasMemory = false }
    fun memoryRecall() { display = fmt(memory); shouldResetDisplay = true }
    fun memoryAdd() { memory += displayValue; hasMemory = true }
    fun memorySubtract() { memory -= displayValue; hasMemory = memory != 0.0 }

    // MARK: - History

    fun recallHistory(entry: HistoryEntry) {
        display = entry.result
        shouldResetDisplay = true
        leftOperand = null
        pendingOperator = null
        expression = entry.expression
        justEvaluated = false
    }

    fun clearHistory() { history.clear() }

    /** Restore persisted history (newest-first ordering preserved). */
    fun loadHistory(entries: List<HistoryEntry>) {
        history.clear()
        history.addAll(entries.take(200))
    }

    fun restoreMemory(value: Double, has: Boolean) {
        memory = value
        hasMemory = has
    }

    // MARK: - Private

    private fun compute(a: Double, op: String, b: Double): Double = when (op) {
        "+" -> a + b
        "−" -> a - b
        "×" -> a * b
        "÷" -> if (b == 0.0) (if (a >= 0) Double.POSITIVE_INFINITY else Double.NEGATIVE_INFINITY) else a / b
        "xʸ" -> a.pow(b)
        else -> b
    }

    private fun factorial(n: Int): Double {
        if (n < 0) return Double.NaN
        if (n > 20) return Double.POSITIVE_INFINITY
        return if (n == 0) 1.0 else n.toDouble() * factorial(n - 1)
    }

    private fun addToHistory(expression: String, result: String) {
        if (result == "Error") return
        history.add(0, HistoryEntry(expression = expression, result = result))
        if (history.size > 200) history.removeAt(history.size - 1)
    }
}
